### ISSUE TITLE
Android non-root user can't launch games when changing download path (workaround available)

### ISSUE THREAD

**Issue by Cee123** [zach-morris/plugin.program.iagl#36]
On Android games won't launch if I choose a different rom path from default. But as soon as I set path back to the default they work again.

This doesn't happen on the PC version of kodi.

**Comment by zach-morris** [comment]
I'd need a debug log to know the issue.  The opposite issue is actually documented here:
https://github.com/zach-morris/plugin.program.iagl/issues/1

**Comment by Philps** [comment]
I have no problems with alternative Rom Path [Kodi18][armv7][FireTv]. As long as the alternate RomPath is changed to a path where Kodi has permissions eg. SDCARD1/Android/obb/org.xbmc.kodi/.. or SDCARD1/Android/data/xbmc.kodi/..

**Comment by Cee123** [comment]
I noticed the problem on an Android Q Box/TV Box. For some reason it wouldn't let me isolate that section of the log file.

**Comment by Cee123** [comment]
It could be a permission prob

**Comment by zach-morris** [comment]
Yes, I'd need a debug log.  You can put it into a hastebin and link it

**Comment by Cee123** [comment]
Alright, I've done that. I've isolated the section of the log file where I was testing IAGL, pasted into hastebin, and here's the link.

https://hastebin.com/usipidaqil.sql

**Comment by Cee123** [comment]
I tried to post the full log file, but it says that it exceeds the maximum length.

**Comment by drinfernoo** [comment]
@Cee123 You wouldn't necessarily have to download them each time if you set your cache size larger than zero.

**Comment by zach-morris** [comment]
Thanks, I see the error but not sure whats causing it.  Can you tell me what the path you're entering is for the alternate download path?

**Comment by zach-morris** [comment]
I also see in your log:
```
NOTICE: IAGL:  File saved to location /storage/sdcard1/Roms/gb/Super Mario Land (World).zip, file size -624556341690293936, file crc UNKNOWN
DEBUG: IAGL:  Checking downloaded file /storage/sdcard1/Roms/gb/Super Mario Land (World).zip validity
```
The issue there is that Kodi does not have permissions to that location, so it can't save a file there.  You need to create a location that Kodi has write permission to or root Kodi so it has permission everywhere.

**Comment by Cee123** [comment]
@drinfernoo How do I change the cache size?

**Comment by Cee123** [comment]
@zach-morris The alternative storage path is /storage/sdcard1/Roms/ basically storing on the SD card. Although the same thing happens when I try an alternative storage location anywhere.

**Comment by Cee123** [comment]
@zach-morris How do you change the permissions on Kodi? I don't think it asks for permissions like other applications do.

**Comment by drinfernoo** [comment]
@Cee123 There's a setting on the first page of settings to change the cache size.

**Comment by Cee123** [comment]
@drinfernoo Thanks, I'll try that see if it helps.

**Comment by Cee123** [comment]
@drinfernoo Under first page of the settings? There are different sections under settings. Not sure which one you mean. Player? System? Display settings? Media?

I've checked all of these, still can't find where to change cache size on kodi 18.

**Comment by drinfernoo** [comment]
@Cee123 On the first page of IAGL's settings, you can change the size of it's cache, making it keep more than one ROM downloaded at a time.

**Comment by Cee123** [comment]
@drinfernoo Oh OK. Sorry I thought you meant Kodi settings. I didn't know that there was an option under iagl settings. Thanks heaps for that I'll try that then.

**Comment by Cee123** [comment]
@drinfernoo Thanks heaps! I changed the cache size to 500 mb as most of the 16gb  space is full. Now I don't have to keep downloading them each time. But they still won't load when I change the default path.

**Comment by zach-morris** [comment]
The issue is the same described in #1, where there is some type of permission error:

```
19:43:22.358 T:11077  NOTICE: IAGL:  Attempting to download file ...GoldenEye%20%28USA%29.zip
19:43:22.358 T:11077  NOTICE: IAGL:  Saving file to /storage/sdcard1/Roms/n64/007 - GoldenEye (USA).zip
...
19:43:46.182 T:11077  NOTICE: IAGL:  File saved to location /storage/sdcard1/Roms/n64/007 - GoldenEye (USA).zip, file size 219043332128, file crc UNKNOWN
19:43:46.213 T:11077   DEBUG: IAGL:  Checking downloaded file /storage/sdcard1/Roms/n64/007 - GoldenEye (USA).zip validity
19:43:46.213 T:11077   DEBUG: IAGL:  File validity check passed.
19:43:46.213 T:11077   DEBUG: IAGL:  Post Process file /storage/sdcard1/Roms/n64/007 - GoldenEye (USA).zip - unzip (builtin)
19:43:46.274 T:9514   DEBUG: CZipManager::GetZipList: failed to stat file zip://%2fstorage%2fsdcard1%2fRoms%2fn64%2f007%20-%20GoldenEye%20(USA).zip/
```

It's a Kodi issue at it's root (see [here](https://github.com/xbmc/xbmc/issues/14690)), and I don't have a workaround currently.  This may potentially be resolved with vfs.libarchive which is a binary addon that can unzip files (rather than using the python implementation which chokes in this case), so I'll look into that.

**Comment by Cee123** [comment]
@zach-morris Oh right, that's interesting. I remember also reading something about trying to change permissions within Kodi on Android and there was nothing but a bunch of people saying it wasn't possible as the current Android developer has left the project, and there wont be any new features anytime soon.

**Comment by Cee123** [comment]
@zach-morris Nice. Hopefully that binary addon can fix the issue in the near future.

**Comment by drinfernoo** [comment]
Do you have any more information on that vfs.libarchive add-on? I'm interested in using it as well, especially if it can handle Unicode characters.

**Comment by zach-morris** [comment]
It should be available in the official repository (although I can't guarantee it's available for every OS).  It's a new binary addon specifically to deal with archive filetypes .7z,.tar.gz,.tar.bz2,.tar.xz,.zip,.rar,.tgz,.tbz2,.gz,.bz2,.xz

**Comment by zach-morris** [comment]
This should work now.  If possible please try 2.1.0 and let me know if the issue persists

**Comment by Cee123** [comment]
Awesome. That's great. I'll try it out see if the issue is resolved.

**Comment by Cee123** [comment]
Strangely enough, the issue seems to still be there. Could be the Android version (5.0) or the box itself.

**Comment by zach-morris** [comment]
OK, some things to try:
Uninstall IAGL, then reinstall
Install vfs.libarchive if it's available for your box

**Comment by drinfernoo** [comment]
@Cee123 You need to change the download location to somewhere that Kodi has access to, which unfortunately, is unlikely to be on your SD card. I'm not sure in what version of Android it changed, but apps can now only write in folder where they have permissions.

**Comment by Cee123** [comment]
@zach-morris I tried to install that plugin but it kept saying it wasnt available on Android which is strange.

**Comment by Cee123** [comment]
@drinfernoo Oh OK, that's what I thought. Seems to work with folders on the device itself but the only problem is the limited space. You are right it works with permissions with different folders on the device but not the SD card or anything external.

**Comment by Cee123** [comment]
Well I installed the plugins for archive support and for RAR support. Not sure exactly what vfs.libarchive is or how to install it on Kodi. And that didn't seem to make a difference. Same thing happens (nothing) when directing the download folder to the roms folder on the SD card.

**Comment by zach-morris** [comment]
VFS.libarchive is a kodi binary addon and should be in the official repository.  It allows a (non-python) method to work with archive files from Kodi, which IAGL now uses if it's installed.  I don't believe it's available for all OS's (yet) though.  See image below.  I'm pretty convinced this is a permissions issue in Android, and somehow permission needs to be granted to Kodi (or all applications) to write to your external storage.  Just googling 'android enable write access to external storage' provides some examples on how to do this, but I think it's really going to vary depending on your OS version, hardware, current permission level, etc.  From the addon, there really isn't anything I can do to elevate permissions.

<img width="350" alt="vfs.libarchive" src="https://i.imgur.com/gQfxuOy.png">

**Comment by Cee123** [comment]
Thanks heaps Zach for all your hard work and assistance. 

I'll try that external storage path. Already tried that Archive Support addon and that didn't work. It could be something to do with the earlier version of Android (5.0). And if that doesn't work either, then no problem, no hard feelings, I'll just set and keep the rom path as the default path which works for me.

**Comment by zach-morris** [comment]
I'd be interested to know if the external storage path method works.  I've already updated the FAQ hoping it's a true statement.

**Comment by Philps** [comment]
So one workaround might be, to install Kodi on the SD Card 
( this work on Amazon Fire TV Gen2, by using move app to SD and this will create the .kodi structure. )
What i wanted to try next, to move the App back and see if i can use the created StorePath which will look as follws on AFTV:

/storage/sdcard1/Android/data/org.xbmc.kodi/files/
User: u0_a14 ( some AFTV specific)
Group: sdcard_r
root and owner/groupd as full access. 

However, i am missing the context menu in IAGL 2.1.1 to select the download path? I will go back to 2.1.0 and check if its a user error or legit

**Comment by Philps** [comment]
> 
> 
> I have no problems with alternative Rom Path [Kodi18][armv7][FireTv]. As long as the alternate RomPath is changed to a path where Kodi has permissions eg. SDCARD1/Android/obb/org.xbmc.kodi/.. or SDCARD1/Android/data/xbmc.kodi/..

has anything been changed from 2.1.0 to 2.1.1? as i am no longer able to download to external SDCARD1

**Comment by Philps** [comment]
> 2019-05-16 11:06:08.426 T:8475   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.IndexError'>
                                            Error Contents: list index out of range
                                            Traceback (most recent call last):
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 4198, in onClick
                                                self.action_download_only()
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 4222, in action_download_only
                                                download_and_process_success = IAGL_DL.download_and_process_game_files() #Download files
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 3748, in download_and_process_game_files
                                                self.post_process_unarchive_files(self.current_saved_files[ii],self.current_saved_files_crc[ii])
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 3196, in post_process_unarchive_files
                                                self.post_process_unarchive_files_libarchive(filename_in,crc_in)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 3259, in post_process_unarchive_files_libarchive
                                                xbmc.log(msg='IAGL:  The file %(filename_in)s was unarchived.  First file extracted: %(files_extracted)s'% {'filename_in': filename_in, 'files_extracted':files_extracted[0]}, level=xbmc.LOGDEBUG)
                                            IndexError: list index out of range
                                            -->End of Python script error report<--

**Comment by Philps** [comment]
It seems storing to external SDCARD works but extracting is an issue. (archive support installed)

2019-05-16 11:21:19.349 T:9838   DEBUG: IAGL:  Source URL returned no filesize, current size estimate is 4325
2019-05-16 11:21:19.353 T:9838  NOTICE: IAGL:  File saved to location /storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/Acid Drop (USA).zip, file size 4325, file crc 4B33D48
2019-05-16 11:21:19.364 T:9838   DEBUG: IAGL:  Checking downloaded file /storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/Acid Drop (USA).zip validity
2019-05-16 11:21:19.371 T:9838   DEBUG: IAGL:  File validity check passed.
2019-05-16 11:21:19.371 T:9838   DEBUG: IAGL:  Post Process file /storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/Acid Drop (USA).zip - unarchive (vfs.libarchive)
2019-05-16 11:21:19.394 T:9838   DEBUG: IAGL:  Extracted file Acid Drop (USA).a26 from archive /storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/Acid Drop (USA).zip
2019-05-16 11:21:19.395 T:9838   DEBUG: IAGL:  The file /storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/Acid Drop (USA).zip was unarchived.  First file extracted: /storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/Acid Drop (USA).a26


SDCARD (external)
2019-05-16 11:22:20.080 T:9957  NOTICE: IAGL:  File saved to location /storage/sdcard1/Android/obb/org.xbmc.kodi/Acid Drop (USA).zip, file size 8242003386457607239, file crc UNKNOWN
2019-05-16 11:22:20.092 T:9957   DEBUG: IAGL:  Checking downloaded file /storage/sdcard1/Android/obb/org.xbmc.kodi/Acid Drop (USA).zip validity
2019-05-16 11:22:20.092 T:9957   DEBUG: IAGL:  File validity check passed.
2019-05-16 11:22:20.093 T:9957   DEBUG: IAGL:  Post Process file /storage/sdcard1/Android/obb/org.xbmc.kodi/Acid Drop (USA).zip - unarchive (vfs.libarchive)
2019-05-16 11:22:20.095 T:9957   ERROR: GetDirectory - Error getting archive://%2fstorage%2fsdcard1%2fAndroid%2fobb%2forg.xbmc.kodi%2fAcid%20Drop%20(USA).zip/
2019-05-16 11:22:20.105 T:9957   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.IndexError'>
                                            Error Contents: list index out of range
                                            Traceback (most recent call last):
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 4198, in onClick
                                                self.action_download_only()
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 4222, in action_download_only
                                                download_and_process_success = IAGL_DL.download_and_process_game_files() #Download files
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 3748, in download_and_process_game_files
                                                self.post_process_unarchive_files(self.current_saved_files[ii],self.current_saved_files_crc[ii])
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 3196, in post_process_unarchive_files
                                                self.post_process_unarchive_files_libarchive(filename_in,crc_in)
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.program.iagl/resources/lib/main.py", line 3259, in post_process_unarchive_files_libarchive
                                                xbmc.log(msg='IAGL:  The file %(filename_in)s was unarchived.  First file extracted: %(files_extracted)s'% {'filename_in': filename_in, 'files_extracted':files_extracted[0]}, level=xbmc.LOGDEBUG)
                                            IndexError: list index out of range
                                            -->End of Python script error report<--

**Comment by Philps** [comment]
Unfortunately i didnt check permission when i hit the failure bute wiped off kodi and all previously created folders on the SDCard. I assume that the issue is simply due to permission and may break when kodi is updated ( different User? when installing a new Kodi APK )

However i tested a fresh Kodi 18.2 install
- IAGL 2.1.0 -> works
- IAGL 2.1.0 + VFS -> works
- IAGL 2.1.1 + VFS -> Works

Folder Permissions on FireTV Gen2 for SDCards - New install Kodi 18.2


lrwxrwxrwx root     root              2019-05-16 11:56 sdcard0 -> /storage/emulated/legacy
drwxrwx--x root     sdcard_r          2019-05-16 11:57 sdcard1


sdcard0
--------
1|shell@sloane:/storage $ ll /storage/sdcard0/
drwxrwx--x root     sdcard_r          2017-11-29 16:49 Android

shell@sloane:/storage $ ll /storage/sdcard0/Android/
drwxrwx--x root     sdcard_r          2019-05-16 12:22 data
drwxrwx--x root     sdcard_r          2017-11-29 16:49 obb

shell@sloane:/storage $ ll /storage/sdcard0/Android/data/
drwxrwx--- u0_a18   sdcard_r          2019-05-16 09:46 com.retroarch
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 org.xbmc.kodi

shell@sloane:/storage $ ll /storage/sdcard0/Android/data/org.xbmc.kodi/
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 files

shell@sloane:/storage $ ls -lisa /storage/sdcard0/Android/data/org.xbmc.kodi/files
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 .kodi

orage/sdcard0/Android/data/org.xbmc.kodi/files/.kodi/
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:24 addons
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 media
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 system
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 temp
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 userdata

sdcard1
--------

shell@sloane:/storage $ ll /storage/sdcard1
drwxrwx--x root     sdcard_r          2019-05-16 12:22 Android

shell@sloane:/storage $ ll /storage/sdcard1/Android/
drwxrwx--x root     sdcard_r          2019-05-16 12:22 data
drwxrwx--x root     sdcard_r          2019-05-16 12:22 obb

shell@sloane:/storage $ ll /storage/sdcard1/Android/data/
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 org.xbmc.kodi

shell@sloane:/storage $ ll /storage/sdcard1/Android/data/org.xbmc.kodi/
drwxrwx--- u0_a14   sdcard_r          2019-05-16 12:22 files

**Comment by zach-morris** [comment]
@Philps , the errors you note I think are resolved in v3.0.  The unarchiving functions are completely rewritten and so long as you have read/write access to the location, it should work.

The requirements in android for external download paths I've updated in the FAQ [here](https://github.com/zach-morris/plugin.program.iagl/wiki/5.--FAQ#15-im-on-android-and-i-cant-change-the-download-path-for-games), and I've added a warning dialog when an android user changes the download path to alert them that they need to know what they're doing.
